### PR TITLE
feat(fw): New `with_all` markers for precompiles, create opcode + other fixes

### PR DESCRIPTION
**Pull request by marioevz**
## 🗒️ Description

This PR implements features useful for EIP-7702 tests.

### `with_all_system_contracts` marker
Adds `with_all_system_contracts` marker and `system_contracts` method to forks, which helps parametrize tests with all contracts that affect the chain on a system level.

### `with_all_create_opcodes` marker
Adds `with_all_create_opcodes` marker and `create_opcodes` method to forks, to parametrize tests with all contract creating opcodes that a given fork supports, along with the appropriate `EVMCodeType` (Legacy or EOF V1).

### Lambda function parameter for `with_all_*` markers
Allow a lambda function filter to be passed as a single parameter to any `with_all_*` marker in order to filter-out parametrization values.

```python
import pytest

@pytest.mark.with_all_tx_types(lambda tx_type: tx_type != 3)
@pytest.mark.valid_from("London")
def test_something_with_all_tx_types(tx_type: int):
    pass
```

In this example, the test will be parametrized with all transaction types except for Type-3.

This is preferred to using `pytest.mark.parametrize` and selecting the values directly because in future forks new types will be added, and even when the marked test fails due to the new transaction not being immediately compatible, it will prompt a proper test update.

### `parametrize_with_defaults` decorator function

`pytest.mark.parametrize` replacement that allows to specify parameter names with default
    values for each parameter.

Example usage:
```python
@parametrize_with_defaults(
    parameter_1=0,  # Default value for parameter_1 is zero
    parameter_2='default',  # Default value for parameter_2 is 'default' string
    cases=[
        dict(parameter_1=1, id='test_1'),
        dict(parameter_2='custom', id='test_2'),
    ],
)
def test(parameter_1, parameter_2):
    ...
```

### Other Fixes
- Fix `compute_create_address` to not use the `sender.nonce` if `nonce` argument is provided.
- Implement EOF V1 `Container.Init` that returns a simple init container which deploys the specified `deployed_container`.
- Add `stop` boolean parameter to `CodeGasMeasure` in order to skip adding `Op.STOP` at the end of the returned bytecode.
- Export `Bytes` from `ethereum_test_tools`.
- Fix decimal values printing on `Storage.KeyValueMismatch`.
- Allow `storage` to be set on `pre.fund_eoa`

## 🔗 Related Issues
None

## ✅ Checklist
- [x] All: Set appropriate labels for the changes.
- [x] All: Considered squashing commits to improve commit history.
- [ ] All: Added an entry to [CHANGELOG.md](/ethereum/execution-spec-tests/blob/main/docs/CHANGELOG.md).
- [x] All: Considered updating the online docs in the [./docs/](/ethereum/execution-spec-tests/blob/main/docs/) directory.
- [x] Tests: All converted JSON/YML tests from [ethereum/tests](/ethereum/tests) have been added to [converted-ethereum-tests.txt](/ethereum/execution-spec-tests/blob/main/converted-ethereum-tests.txt).
- [x] Tests: A PR with removal of converted JSON/YML tests from [ethereum/tests](/ethereum/tests) have been opened.
- [x] Tests: Included the type and version of evm t8n tool used to locally execute test cases:  e.g., ref with commit hash or geth 1.13.1-stable-3f40e65.
- [x] Tests: Ran `mkdocs serve` locally and verified the auto-generated docs for new tests in the [Test Case Reference](https://ethereum.github.io/execution-spec-tests/main/tests/) are correctly formatted.

### CODE REVIEW

**Comment by marioevz** [comment]
An alternative to `named_pytest_param` could be the following:
```python
@named_parametrize(
    signer_type=SignerType.SINGLE_SIGNER,
    authorization_invalidity_type=AuthorizationInvalidityType.NONE,
    authorizations_count=1,
    chain_id_type=ChainIDType.GENERIC,
    authorize_to_address=AddressType.EMPTY_ACCOUNT,
    access_list_case=AccessListType.EMPTY,
    self_sponsored=False,
    authority_type=AddressType.EMPTY_ACCOUNT,
    data=b"",
    cases=[
        dict(authorizations_count=2, id="two_authorizations", marks=[pytest.mark.valid_from("Berlin")]),
        ...
    ],
)
def test_gas(
    blockchain_test: BlockchainTestFiller,
    signer_type: SignerType,
    authorization_invalidity_type: AuthorizationInvalidityType,
    authorizations_count: int,
    chain_id_type: ChainIDType,
    authorize_to_address: AddressType,
    access_list_case: AccessListType,
    self_sponsored: bool,
    authority_type: AddressType,
    data: bytes,
):
    ... 
```

@spencer-tb @danceratopz

File: src/pytest_plugins/forks/forks.py [hunk 1]
```diff
@@ -125,12 +125,47 @@ class CovariantDescriptor:
...
+
+        I.e. if the marker has an argument, the argument is interpreted as a lambda function
+        that filters the values.
+        """
+        marker = self.get_marker(metafunc)
+        assert marker is not None
+        if len(marker.args) == 0:
+            return values
+        assert len(marker.args) == 1, f"Invalid number of arguments for marker {self.marker_name}"
+        filter_func = marker.args[0]
```
Comment (danceratopz) [inline src/pytest_plugins/forks/forks.py:159]:
```suggestion
        filter_func = marker.args[0]
        assert isinstance(filter_func, FunctionType), "First agument to marker must be of FunctionType"
```
If you want to add this check, don't forget:
```python
from types import FunctionType
```

**Review by danceratopz — COMMENTED** [review]
This is fantastic. It's so nice how cleanly the domain logic is separated in `src/ethereum_test_forks/forks/forks.py`.

I like the implementation as-is, but I do think we need to add comprehensive testing for it. I'd probably do this on a high-level using `pytester` as done for other forks plugin markers (e.g., src/pytest_plugins/forks/tests/test_forks.py), what do you think? I'm happy to help add this!

**Comment by danceratopz** [comment]
@marioevz And thanks for the documentation! Really nice, I added a couple of commits that mainly make the examples a more complete starting point for a test if copy-pasted. Hope you like it, feel free to revert if you find it too noisy.

**Review by spencer-tb — APPROVED** [review]
Really nice PR!

The only suggestion I have is the naming for the new decorator.

I find `@default_params` or `@default_parameters` more intuitive than `@named_parameterize`!

**Comment by marioevz** [comment]
> I find `@default_params` or `@default_parameters` more intuitive than `@named_parameterize`!

I feel `params` could be confused with `pytest.param`?
It's still a replacement for `parametrize`, so I'd like for it to include word `parametrize` if possible.
Open to other suggestions :)

**Comment by spencer-tb** [comment]
> > I find `@default_params` or `@default_parameters` more intuitive than `@named_parameterize`!
> 
> I feel `params` could be confused with `pytest.param`? It's still a replacement for `parametrize`, so I'd like for it to include word `parametrize` if possible. Open to other suggestions :)

What about `default_parameterize`? I think I like the "default" keyword here as we are defining the defaults to parameterize with!

Happy with `named` nonetheless :)

**Comment by danceratopz** [comment]
@marioevz What do you think about extending the interface to the `with_all_...` markers with these keyword arguments:

```python
@pytest.mark.with_all_call_opcodes(
    param_name="call_opcode,evm_code_type",  # obligatory argument?
    selector=lambda call_opcode, evm_code_type: call_opcode != Op.CALL,  # optional; default=True
    marks=[  # optional; default=[]
        with_all_mark(
            condition=lambda call_opcode, evm_code_type: call_opcode == Op.EXTDELEGATECALL,
            # alternative syntax proposal for condition:
            # condition="call_opcode==Op.EXTDELEGATECALL and evm_code_type==EVMCodeType.EOF_V1",
            mark=pytest.mark.xfail(reason="EXTDELEGATECALL not implemented"),
        )
    ],
)
```

The motivation behind `param_name` is that the casual reader of a test has no idea which parameter(s!) are even getting parametrized (which parametrizing fixture should he search for), see for example this test (within the context of the module, the `with_all_call_opcodes` overwrites the fixture defined on L100):
https://github.com/ethereum/execution-spec-tests/blob/632d151ea8a71d09a3a0acbdb85d97fa18c8456b/tests/cancun/eip4844_blobs/test_point_evaluation_precompile.py#L488-L523

The test implementer should at least be aware of which parameters he has available to him and can choose to name them here. We can recommend that implementers always use the same names for consistency.

These names can be re-used in the `selector` lambda and the `condition` to apply marks.

If you like the idea, I'm happy to extend your implementation to support this and add testing.

**Comment by marioevz** [comment]
> @marioevz What do you think about extending the interface to the `with_all_...` markers with these keyword arguments:
> 
> ```python
> @pytest.mark.with_all_call_opcodes(
>     param_name="call_opcode,evm_code_type",  # obligatory argument?
>     selector=lambda call_opcode, evm_code_type: call_opcode != Op.CALL,  # optional; default=True
>     marks=[  # optional; default=[]
>         with_all_mark(
>             condition=lambda call_opcode, evm_code_type: call_opcode == Op.EXTDELEGATECALL,
>             # alternative syntax proposal for condition:
>             # condition="call_opcode==Op.EXTDELEGATECALL and evm_code_type==EVMCodeType.EOF_V1",
>             mark=pytest.mark.xfail(reason="EXTDELEGATECALL not implemented"),
>         )
>     ],
> )
> ```
> 
> The motivation behind `param_name` is that the casual reader of a test has no idea which parameter(s!) are even getting parametrized (which parametrizing fixture should he search for), see for example this test (within the context of the module, the `with_all_call_opcodes` overwrites the fixture defined on L100):
> 
> https://github.com/ethereum/execution-spec-tests/blob/632d151ea8a71d09a3a0acbdb85d97fa18c8456b/tests/cancun/eip4844_blobs/test_point_evaluation_precompile.py#L488-L523
> 
> The test implementer should at least be aware of which parameters he has available to him and can choose to name them here. We can recommend that implementers always use the same names for consistency.
> 
> These names can be re-used in the `selector` lambda and the `condition` to apply marks.
> 
> If you like the idea, I'm happy to extend your implementation to support this and add testing.

I rather just put a comment in front of the marker
```python
@pytest.mark.with_all_call_opcodes()  # Parametrizes "call_opcode" parameter
```
than having more boilerplate code and allowing the tester to put any name in there.

**Comment by danceratopz** [comment]
> I rather just put a comment in front of the marker
> 
> ```python
> @pytest.mark.with_all_call_opcodes()  # Parametrizes "call_opcode" parameter
> ```
> 
> than having more boilerplate code and allowing the tester to put any name in there.

Ok, what about the marks?

**Comment by marioevz** [comment]
> > I rather just put a comment in front of the marker
> > ```python
> > @pytest.mark.with_all_call_opcodes()  # Parametrizes "call_opcode" parameter
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > than having more boilerplate code and allowing the tester to put any name in there.
> 
> Ok, what about the marks?

Yes this one would be nice, I think that could go in its own separate PR though because this already introduces too many stuff and it's required for EIP-7702 tests.

**Pull request merged**